lib/basedriver/capabilities.js

```
@@ -0,0 +1,111 @@
+                                          
+  if (validationErrors) {
+    for (let [attribute, reasons] of _.toPairs(validationErrors)) {
+      for (let reason of reasons) {
+        message += ` ${attribute} ${reason},`;
+      }
+    }
+    message = `${message.slice(0, -1)}.`;
+    throw new Error(message);
+  }
+
+  // Return caps
+  return caps;
+}
+
+// Parse capabilities (based on https://www.w3.org/TR/webdriver/#processing-capabilities)
+function parseCaps (caps, constraints = {}, shouldValidateCaps = true) {
+  if (!_.isObject(caps)) {
+    throw new Error('The capabilities argument was not valid for the following reason(s): "capabilities" must be a JSON object.');
+  }
+
+  let {alwaysMatch: requiredCaps, firstMatch: allFirstMatchCaps = []} = caps; // jshint ignore:line
+
+  if (!_.isObject(requiredCaps)) {
+    requiredCaps = {}; 
+  }
+
+  requiredCaps = validateCaps(requiredCaps, constraints);
+
+  // Reject 'firstMatch' argument if it's not an array (see spec #3.2)
+  if (!_.isArray(allFirstMatchCaps)) {
+    throw new Error('The capabilities.firstMatch argument was not valid for the following reason(s): "capabilities.firstMatch" must be a JSON array or undefined');
+  }
+
+  // Validate the firstMatch capabilities one-by one (see spec #5)
+  let validatedFirstMatchCaps = allFirstMatchCaps.map((firstMatchCaps, index) => {
+    try {
+      return shouldValidateCaps ? validateCaps(firstMatchCaps, constraints) : firstMatchCaps;
+    } catch (e) {
+      throw new Error(`The capabilities.firstMatch argument was not valid for the following reason: "capabilities.firstMatch[${index}] ${e.message}`);
+    }
+  });
+
+  // Try to match requiredCaps with one of the validated firstMatch capabilities (see spec #6)
+  let matchedCaps = null;
+  for (let firstMatchCaps of validatedFirstMatchCaps) {
+    try {
+      let matchedCaps = mergeCaps(requiredCaps, firstMatchCaps);
+      if (matchedCaps) {
+        break;
+      }
+    } catch (ign) { }
```

If we're ignoring this error here, why even throw it on line 13?